website_canonical_url/models/website.py

```diff
@@ -0,0 +1,35 @@
+# -*- coding: utf-8 -*-
+# Copyright initOS GmbH 2016
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import models, api
+from odoo.http import request
+from urlparse import urlparse, urlunparse
+
+
+class Website(models.Model):
+    _inherit = 'website'
+
+    @api.multi
+    def get_canonical_url(self, req=None):
+        canonical_url = None
+        if req is None:  # pragma: no cover
+            req = request
+        if req and req.httprequest.path:
+            lang = self.env.lang
+            if lang == request.website.default_lang_code:
+                canonical_url = req.httprequest.path
+            else:
+                url_parts = urlparse(req.httprequest.path)
```

Comment:
@Yajo @zeroheure I'm refactoring and writing tests but I don't get some things here and current implementations seems broken to me:

1. language code is added always when is not default, but is already in the path (and even better is cleaned up in `urlparse(req.httprequest.path).path`) 
2. querystring params are not stripped out which is kind of wrong AFAIK

Am I wrong?

Reply:
to me it should be something like:
```python
parsed = urlparse(req.httprequest.path)
canonical_url = parsed.path
if lang != req.website.default_lang_code:
    for lang, __ in self.get_languages():
        if canonical_url.startswith(lang):
            canonical_url = canonical_url[len(lang):]
            break
```
kind of....